src/Feature/Editor/EditorConfiguration.re

```diff
@@ -0,0 +1,79 @@
+open Revery;
+open Oni_Core;
+open Config.Schema;
+
+module CustomDecoders: {
+  let whitespace:
+    Config.Schema.decoder(option([ | `All | `Boundary | `None]));
+  let lineNumbers:
+    Config.Schema.decoder(option([ | `On | `Relative | `Off]));
+  let time: Config.Schema.decoder(option(Time.t));
+} = {
+  open Json.Decode;
+
+  let whitespace =
+    custom(
+      string
+      |> map(
+           fun
+           | "none" => `None
+           | "boundary" => `Boundary
+           | "all"
+           | _ => `All,
+         ),
+    );
+
+  let lineNumbers =
+    custom(
+      string
+      |> map(
+           fun
+           | "off" => `Off
+           | "relative" => `Relative
+           | "on"
+           | _ => `On,
+         ),
+    );
+
+  let time = custom(int |> map(Time.ms));
+};
+
+open CustomDecoders;
+
+let detectIndentation =
+  setting("editor.detectIndentation", bool |> default(true));
+let fontFamily =
+  setting("editor.fontFamily", string |> default("FiraCode-Regular.ttf"));
+let fontSize = setting("editor.fontSize", int |> default(14));
+let largeFileOptimization =
+  setting("editor.largeFileOptimizations", bool |> default(true));
+let highlightActiveIndentGuide =
+  setting("editor.highlightActiveIndentGuide", bool |> default(true));
+let indentSize = setting("editor.indentSize", int |> default(4));
+let insertSpaces = setting("editor.insertSpaces", bool |> default(false));
+let lineNumbers = setting("editor.lineNumbers", lineNumbers |> default(`On));
+let matchBrackets = setting("editor.matchBrackets", bool |> default(true));
```

Comment:
I can't imagine any case where we wouldn't want a default, even if the default is just like `""` for a path etc, or a `None` equivalent for some options.

If its easier/simpler code and it forces us to always set a sane default, I can't see a downside.

Reply:
Yeah, that's what eventually started dawning on me as well.

Reply:
Makes sense to me to always have a default as well 👍

Reply:
I really like having the settings completely described in this way! 

I could see in the future - extending this to a friendly name + a description. This could be used a couple of ways:
- Auto-generate our documentation
- Provide completion / suggestions when editing our configuration.

Very cool @glennsl

Reply:
Definitely. That would require them to be collected centrally somehow though. We could do that impurely by registering them in a global, otherwise we'd need to have a list of contributions.  The problem with the latter is that these must be parameterized by the value for the getter to be able to return the value appropriately typed. We don't need that centrally though, so we could either hide it using an existential type or just extract the non-parameterized data. Either way, it requires making another function call for each item before putting it in the contributions list.